mlx/ops.cpp

```diff
@@ -3273,6 +3273,26 @@ array tensordot(
   return reshape(matmul(x, y, s), rshape, s);
 }
 
+array dot(const array& a, const array& b, StreamOrDevice s /* = {} */) {
+  if (a.ndim() == 0 || b.ndim() == 0) {
+    return multiply(a, b, s);
+  }
+  if (a.ndim() == 1 && b.ndim() == 1) {
+    return inner(a, b, s);
+  }
+  if (a.ndim() == 2 && b.ndim() == 2) {
+    return matmul(a, b, s);
+  }
+  if (b.ndim() == 1) {
+    return tensordot(a, b, {{-1}, {-1}}, s);
+  }
```

Comment:
Do you want a shape check here as well like you have below?

Reply:
tensordot does a shape check, can add it here or leave it to tensordot to check and remove the other check under dot before calling dot.

The only side effect is that the error message will show `[tensordot]` instead of dot if leaving it to tensordot to do the check